llvm/include/llvm/Support/SYCLRTShared.h

```diff
@@ -0,0 +1,23 @@
+//=- SYCLRTShared.h - Shared definition between llvm tools and SYCL runtime -=//
```

Comment:
This seems OK to me.

Reply:
Hi, @kbobrovs 
I have added the comments in SYCLRTShared.h.

Reply:
@kbobrovs, thanks for pinging me and sorry for late feedback.

I don't fully understand the design of this "linking on demand" solution (it would be great to extend https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md document), but this concerns me.
It seems like we bake into LLVM compiler framework the information about Intel specific OpenCL extensions. @andykaylor, do you think this design can be accepted by the LLVM community? I think community prefers scaleable solutions (e.g. framework for linking arbitrary libraries), whereas current implementation is able to address very specific SYCL use case.

It would be great to get a community feedback on this approach through RFC.

Reply:
> It seems like we bake into LLVM compiler framework the information about Intel specific OpenCL extensions.
@bader, under  "Intel specific OpenCL extensions" do you mean SYCL as a whole or just the "linking on demand part"?
If the former (SYCL), then in my view we already have baked a lot of SYCL into LLVM :) (for good reason) - take clang FE, for example, offload tools - bundler, wrapper, SYCL specific passes in LLVM itself. Maybe the location "llvm/Support/SYCLRTShared.h" is not ideal and there could be new folder with "SYCL" in it to underline SYCL specifics?

If the latter - then SYCLRTShared.h is not specific to this extension, it serves for sharing some type and string literal definitions between LLVM tools and SYCL runtime.

> It would be great to get a community feedback on this approach through RFC.
Do you think this should be done before this PR can be merged or can be done in parallel?

Reply:
> > It seems like we bake into LLVM compiler framework the information about Intel specific OpenCL extensions.
> 
> @bader, under "Intel specific OpenCL extensions" do you mean SYCL as a whole or just the "linking on demand part"?

cl_intel_devicelib_assert, cl_intel_devicelib_math, cl_intel_devicelib_math_fp64, cl_intel_devicelib_complex, cl_intel_devicelib_complex_fp64 - are Intel specific OpenCL extensions.

> If the former (SYCL), then in my view we already have baked a lot of SYCL into LLVM :) (for good reason) - take clang FE, for example, offload tools - bundler, wrapper, SYCL specific passes in LLVM itself. Maybe the location "llvm/Support/SYCLRTShared.h" is not ideal and there could be new folder with "SYCL" in it to underline SYCL specifics?
> 
> If the latter - then SYCLRTShared.h is not specific to this extension, it serves for sharing some type and string literal definitions between LLVM tools and SYCL runtime.
> 

Can we apply the same approach as regular C++ compiler does? I.e. link with the whole C++ standard library (using LLVM linker) + provide an compiler option to disable [SYCL extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst) and skip linking with this libraries? This should much easier to implement and maintain as it should not require extending LLVM functionality.

> > It would be great to get a community feedback on this approach through RFC.
> 
> Do you think this should be done before this PR can be merged or can be done in parallel?

Ideally the review should happen before the merge, but if there are any time constraints we can merge and re-do later if needed.

Reply:
Hi, @kbobrovs @andykaylor @bader 
This patch is used to fix some perf regression for cases which has "small" kernel execution time and we are required to fix it ASAP recently, could we merge it and continue to work on the RFC to collect community's feedback?
And I think it is valuable to add a header to share information between sycl runtime and llvm-tools. Currently, the header is located in llvm/Support/SYCLRTShared.h and following things are included:
1. Property constant string such as "SYCL/specialization constants".
2. DeviceLibExt enum definition such as "cl_intel_devicelib_***"
We need a careful analysis and discussion on following points:
1.  The header's filename and location?
2.  The namespace used in this header?
3.  Which  definitions or constants can be placed in this file?
4. Can intel-specific things be placed in this header?
Thank you very much.

Reply:
I'm not convinced that this is the right approach to "fix performance regression". Moving some runtime overhead to compile time (e.g. https://github.com/intel/llvm/pull/1398) seems to be a better option.
What is the root cause of the regression? Please, add a problem statement to the PR description.